File: frontend/src/lib/types/analytics.ts
export interface UsageTotals {
  cache_write_tokens?: number | null;
  runs: number;
  known_cost_usd: string;
  cost_usd: string | null;
  cost_complete: boolean;
  unknown_cost_runs: number;
  incomplete_usage_runs: number;
  reserved_usd: string;
  failed_spend_usd: string;
  compaction_spend_usd: string;
  input_tokens: number | null;
  output_tokens: number | null;
  cache_read_tokens: number | null;
  reasoning_tokens: number | null;
}
export interface UsageGroup extends UsageTotals {
  key: string;
}
export interface AgentEfficiency extends UsageGroup {
  teams?: string[];
  tokens_per_merged_task?: number | null;
  p90_developer_seconds?: number | null;
  display_name: string;
  tasks_merged: number;
  terminal_tasks: number;
  success_rate: number | null;
  median_cost_usd: number | null;
  p90_cost_usd: number | null;
  median_developer_seconds: number | null;
  human_intervention_rate: number | null;
  review_fix_cycles_per_task: number | null;
  models: string[];
  harnesses: string[];
}
export interface AnalyticsDashboard {
  month_totals?: UsageTotals;
  period_costs?: Record<string, UsageTotals>;
  repository_names?: Record<string, string>;
  reliability?: Record<string, number>;
  local_runs?: LocalRun[];
  sampled_at: string;
  days: number;
  totals: UsageTotals;
  agents: AgentEfficiency[];
  models: UsageGroup[];
  run_kinds: UsageGroup[];
  repositories: UsageGroup[];
  merged_tasks: number;
  complete_merged_tasks: number;
  excluded_incomplete_tasks: number;
  cost_per_merged_task_usd: string | null;
  basis: string;
  daily: UsageGroup[];
}
export interface CostForecast {
  metrics?: Record<string, ForecastMetric>;
  predicted_peak_concurrent_memory_bytes?: number | null;
  queue_drain_seconds?: number | null;
  resource_basis?: string;
  forecast_version: string;
  confidence: string;
  sample_count: number;
  estimate: number | null;
  range: { p50: number | null; p90: number | null } | null;
  basis: string;
  queued_tasks?: number;
  days?: number;
}
export interface TaskAnalytics extends Omit<UsageTotals, 'runs'> {
  runs?: RunUsage[];
  local_runs?: LocalRun[];
  wait_seconds_by_reason?: Record<string, number>;
  phase_seconds?: Record<string, number>;
  provider_active_seconds?: number | null;
  provider_wait_seconds?: number | null;
  time_to_first_edit_seconds?: number | null;
  task_id: string;
  title: string;
  status: string;
  developer_active_seconds: number | null;
  wall_seconds: number | null;
  human_or_external_wait_seconds: number;
  phases: Record<string, number>;
  developer_turns: number;
  compactions: number;
  review_cycles: number;
  validation_failures: number;
  peak_memory_bytes: number | null;
  warnings: string[];
  forecasts: {
    forecast_version: string;
    confidence: string;
    sample_count: number;
    estimate: Record<string, unknown>;
    actuals: Record<string, unknown> | null;
  }[];
}

export interface ForecastMetric {
  estimate: number | null;
  range: { p50: number | null; p90: number | null; p75?: number | null };
  sample_count?: number;
  confidence?: string;
}
export interface LocalRun {
  model: string;
  status: string;
  duration_ms: number | null;
  input_tokens: number | null;
  output_tokens: number | null;
  started_at: string;
}
export interface RunUsage {
  id: string;
  model: string;
  harness: string;
  run_kind: string;
  status: string;
  cost: string | null;
  input_tokens: number | null;
  output_tokens: number | null;
  cache_read_tokens: number | null;
  cache_write_tokens: number | null;
  reasoning_tokens: number | null;
  started_at: string;
  finished_at: string | null;
  provider_duration_ms: number | null;
}
export interface ForecastAccuracy {
  metrics: Record<
    string,
    {
      sample_count: number;
      median_absolute_percentage_error: number | null;
      p90_coverage: number | null;
      mean_bias: number | null;
    }
  >;
  basis: string;
}

File: backend/app/interfaces/http/schemas/analytics.py
from datetime import datetime
from decimal import Decimal

from pydantic import Field

from app.interfaces.http.schemas.observability import ExtensibleResponse


class UsageTotalsResponse(ExtensibleResponse):
    known_cost_usd: str
    cost_usd: str | None
    cost_complete: bool
    unknown_cost_runs: int
    incomplete_usage_runs: int
    reserved_usd: str
    failed_spend_usd: str
    compaction_spend_usd: str
    input_tokens: int | None
    output_tokens: int | None
    cache_read_tokens: int | None
    cache_write_tokens: int | None
    reasoning_tokens: int | None


class UsageGroupResponse(UsageTotalsResponse):
    key: str
    runs: int


class AgentEfficiencyResponse(UsageGroupResponse):
    display_name: str
    tasks_merged: int
    terminal_tasks: int
    success_rate: float | None
    median_cost_usd: float | None
    p90_cost_usd: float | None
    median_developer_seconds: float | None
    models: list[str]
    harnesses: list[str]


class AnalyticsDashboardResponse(ExtensibleResponse):
    sampled_at: datetime
    days: int
    totals: UsageTotalsResponse
    month_totals: UsageTotalsResponse
    period_costs: dict[str, UsageTotalsResponse]
    agents: list[AgentEfficiencyResponse]
    models: list[UsageGroupResponse]
    repositories: list[UsageGroupResponse]
    run_kinds: list[UsageGroupResponse]
    daily: list[UsageGroupResponse]
    merged_tasks: int
    complete_merged_tasks: int
    excluded_incomplete_tasks: int
    cost_per_merged_task_usd: str | None
    basis: str


class ForecastRangeResponse(ExtensibleResponse):
    p50: float | None
    p90: float | None


class ForecastMetricResponse(ExtensibleResponse):
    estimate: float | None
    range: ForecastRangeResponse


class CostForecastResponse(ExtensibleResponse):
    forecast_version: str
    confidence: str
    sample_count: int
    estimate: float | None
    range: ForecastRangeResponse | None
    metrics: dict[str, ForecastMetricResponse]
    basis: str


class TaskForecastResponse(ExtensibleResponse):
    forecast_version: str
    confidence: str
    sample_count: int
    estimate: dict[str, ForecastMetricResponse] | None
    range: dict[str, ForecastRangeResponse] | None


class RunUsageResponse(ExtensibleResponse):
    id: str
    model: str
    harness: str | None
    run_kind: str
    status: str
    cost: Decimal | None
    input_tokens: int | None
    output_tokens: int | None
    cache_read_tokens: int | None
    cache_write_tokens: int | None
    reasoning_tokens: int | None
    started_at: datetime
    finished_at: datetime | None
    provider_duration_ms: int | None


class TaskAnalyticsResponse(UsageTotalsResponse):
    task_id: str
    title: str
    status: str
    runs: list[RunUsageResponse]
    forecasts: list[TaskForecastResponse]
    developer_active_seconds: float | None
    wall_seconds: float | None
    human_or_external_wait_seconds: float
    phases: dict[str, float]
    peak_memory_bytes: float | None
    warnings: list[str] = Field(default_factory=list)


class AccuracyMetricResponse(ExtensibleResponse):
    sample_count: int
    median_absolute_percentage_error: float | None
    p90_coverage: float | None
    mean_bias: float | None


class ForecastAccuracyResponse(ExtensibleResponse):
    metrics: dict[str, AccuracyMetricResponse]
    basis: str
